File: homework2/hangman/src/main/java/id1212/wachsler/joel/hangman/client/controller/Controller.java
package id1212.wachsler.joel.hangman.client.controller;

import id1212.wachsler.joel.hangman.client.net.Listener;
import id1212.wachsler.joel.hangman.client.net.ServerConnection;

import java.io.IOException;
import java.util.concurrent.CompletableFuture;

/**
 * Controller for the client.
 */
public class Controller {
  private final ServerConnection serverConnection = new ServerConnection();

  /**
   * @see ServerConnection#addListener(Listener)
   */
  public void addServerConnectionListener(Listener listener) {
    serverConnection.addListener(listener);
  }

  /**
   * @see ServerConnection#connect(String, int)
   */
  public void connect(String host, int port) {
    CompletableFuture.runAsync(() -> serverConnection.connect(host, port));
  }

  /**
   * @see ServerConnection#disconnect()
   */
  public void disconnect() throws IOException {
    serverConnection.disconnect();
  }

  /**
   * @see ServerConnection#isConnected()
   */
  public boolean isConnected() {
    return serverConnection.isConnected();
  }

  /**
   * @see ServerConnection#sendGuess(String)
   */
  public void guess(String guessingWord) {
    // Send async guess
    CompletableFuture.runAsync(() -> serverConnection.sendGuess(guessingWord));
  }

  /**
   * @see ServerConnection#startGame()
   */
  public void startGame() {
    CompletableFuture.runAsync(() -> serverConnection.startGame());
  }
}

File: homework5/client/app/src/main/java/id1212/wachsler/joel/homework5/controller/Controller.kt
package id1212.wachsler.joel.homework5.controller

import id1212.wachsler.joel.homework5.net.ServerConnection
import java.io.IOException
import java.util.concurrent.CompletableFuture

object Controller {

  /**
   * Tries to connect to the server on a separate thread.
   * If something goes wrong the <code>error</code> callback will be called with the error
   * else the <code>success</code> callback will be called.
   */
  fun connect(error: (msg: String) -> Unit, success: () -> Unit) {
    CompletableFuture.runAsync({
      try {
        ServerConnection.connect()
        success()
      } catch (e: IOException) {
        error(e.message!!)
      }
    })
  }

  /**
   * @see ServerConnection#guess
   */
  fun guess(guess: String) {
    CompletableFuture.runAsync({
      ServerConnection.guess(guess)
    })
  }
}
